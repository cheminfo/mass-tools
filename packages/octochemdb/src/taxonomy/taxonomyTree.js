import { taxonomyRanks } from './taxonomyRanks';
/**
 * @description Creates a tree structure from an array of taxonomies.
 *
 * @param {import('./Taxonomy.js').Taxonomy[]} taxonomies - The array of taxonomies to create a tree from.
 * @returns {Object[]} The tree structure.
 */
export function taxonomyTree(taxonomies) {
  const tree = [];

  for (let taxonomy of taxonomies) {
    let current = tree;
    for (let rank of taxonomyRanks) {
      const name = taxonomy[rank];
      if (!name) continue;
      let existing = current.find(
        (node) => node.name === name && node.rank === rank,
      );
      if (!existing) {
        existing = {
          name,
          rank,
          count: 1,
          children: [],
        };

        current.push(existing);
      } else {
        existing.count++;
      }
      current = existing.children;
    }
  }

  for (let branch of tree) {
    cleanEmptyBranches(branch);
  }
  return tree;
}

function cleanEmptyBranches(branch) {
  branch.children.forEach((child) => cleanEmptyBranches(child));

  branch.children = branch.children.filter((child) => {
    // This part is used empty nodes
    if (child.children.length === 0 && child.name === '') {
      return false;
    }
    // This part is used to remove the children in species rank
    if (child.children.length === 0 && child.name !== '') {
      delete child.children;
    }

    return true;
  });
}
