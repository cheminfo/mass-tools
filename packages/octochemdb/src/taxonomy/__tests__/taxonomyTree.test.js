import { taxonomyTree } from '../taxonomyTree';

test('taxonomyTree', () => {
  const taxonomies = [
    {
      superKingdom: 'Eukaryota',
      kingdom: 'Fungi',
      phylum: 'Ascomycota',
      class: 'Saccharomycetes',
      order: 'Saccharomycetales',
      family: 'Debaryomycetaceae',
      genus: 'Candida',
      species: 'Candida albicans',
    },
    {
      superKingdom: 'bacteria',
      phylum: 'Apicomplexa',
      class: 'Aconoidasida',
      order: 'Haemosporida',
      family: 'Plasmodiidae',
      genus: 'Plasmodium',
      species: 'Plasmodium falciparum',
    },
    {
      superKingdom: 'Eukaryota',
      kingdom: 'Fungi',
      phylum: 'Ascomycota',
      class: 'Sordariomycetes',
      order: 'Hypocreales',
      family: 'Stachybotryaceae',
      genus: 'Stachybotrys',
      species: 'Stachybotrys chartarum',
    },
    {
      superKingdom: 'bacteria',
      kingdom: 'Fungi',
      phylum: 'Ascomycota',
      class: 'Sordariomycetes',
      order: 'Hypocreales',
      family: 'Stachybotryaceae',
      genus: 'Stachybotrys',
      species: 'Stachybotrys cylindrospora',
    },
  ];

  let tree = taxonomyTree(taxonomies);
  console.log(JSON.stringify(tree[1], null, 2));
  expect(tree[1].children).toHaveLength(2);
  expect(tree[1].children[0].name).toBe('');
  expect(tree[1].children[1].name).toBe('Fungi');
  expect(tree[0].children[0].children[0].name).toBe('Ascomycota');
  expect(tree[0].children[0].children[0].children).toHaveLength(2);

  expect(tree).toMatchSnapshot();
});
test('only one entry', () => {
  const taxonomies = [
    {
      superKingdom: 'Eukaryota',
    },
  ];

  let tree = taxonomyTree(taxonomies);

  expect(tree[0].name).toBe('Eukaryota');
  expect(tree[0].count).toBe(1);

  expect(tree).toMatchSnapshot();
});
test('multiple undefined fields', () => {
  const taxonomies = [
    {
      kingdom: 'bacteria',
    },
    {
      superKingdom: 'Eukaryota',
    },
    {
      kingdom: 'bacteria',
      phylum: 'Ascomycota',
    },
    {
      kingdom: 'bacteria',
      phylum: 'Ascomycota',
    },
    {
      superKingdom: 'Eukaryota',
      kingdom: 'bacteria',
      phylum: 'Ascomycota',
    },
  ];

  let tree = taxonomyTree(taxonomies);
  expect(tree[1].name).toBe('Eukaryota');
  expect(tree[1].count).toBe(2);
  expect(tree[1].children[0].name).toBe('bacteria');
  expect(tree[1].children[0].count).toBe(1);
  expect(tree[0].children[0].name).toBe('bacteria');
  expect(tree[0].children[0].count).toBe(3);
  expect(tree).toMatchSnapshot();
});
