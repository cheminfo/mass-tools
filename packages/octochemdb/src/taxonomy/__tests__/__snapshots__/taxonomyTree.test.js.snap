// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`multiple undefined fields 1`] = `
[
  {
    "children": [
      {
        "count": 2,
        "name": "Ascomycota",
        "rank": "phylum",
      },
    ],
    "count": 3,
    "name": "bacteria",
    "rank": "kingdom",
  },
  {
    "children": [
      {
        "children": [
          {
            "count": 1,
            "name": "Ascomycota",
            "rank": "phylum",
          },
        ],
        "count": 1,
        "name": "bacteria",
        "rank": "kingdom",
      },
    ],
    "count": 2,
    "name": "Eukaryota",
    "rank": "superKingdom",
  },
]
`;

exports[`only one entry 1`] = `
[
  {
    "children": [],
    "count": 1,
    "name": "Eukaryota",
    "rank": "superKingdom",
  },
]
`;

exports[`taxonomyTree 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [
                          {
                            "children": [
                              {
                                "count": 1,
                                "name": "Candida albicans",
                                "rank": "species",
                              },
                            ],
                            "count": 1,
                            "name": "Candida",
                            "rank": "genus",
                          },
                        ],
                        "count": 1,
                        "name": "Debaryomycetaceae",
                        "rank": "family",
                      },
                    ],
                    "count": 1,
                    "name": "Saccharomycetales",
                    "rank": "order",
                  },
                ],
                "count": 1,
                "name": "Saccharomycetes",
                "rank": "class",
              },
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [
                          {
                            "children": [
                              {
                                "count": 1,
                                "name": "Stachybotrys chartarum",
                                "rank": "species",
                              },
                            ],
                            "count": 1,
                            "name": "Stachybotrys",
                            "rank": "genus",
                          },
                        ],
                        "count": 1,
                        "name": "Stachybotryaceae",
                        "rank": "family",
                      },
                    ],
                    "count": 1,
                    "name": "Hypocreales",
                    "rank": "order",
                  },
                ],
                "count": 1,
                "name": "Sordariomycetes",
                "rank": "class",
              },
            ],
            "count": 2,
            "name": "Ascomycota",
            "rank": "phylum",
          },
        ],
        "count": 2,
        "name": "Fungi",
        "rank": "kingdom",
      },
    ],
    "count": 2,
    "name": "Eukaryota",
    "rank": "superKingdom",
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [
                          {
                            "count": 1,
                            "name": "Plasmodium falciparum",
                            "rank": "species",
                          },
                        ],
                        "count": 1,
                        "name": "Plasmodium",
                        "rank": "genus",
                      },
                    ],
                    "count": 1,
                    "name": "Plasmodiidae",
                    "rank": "family",
                  },
                ],
                "count": 1,
                "name": "Haemosporida",
                "rank": "order",
              },
            ],
            "count": 1,
            "name": "Aconoidasida",
            "rank": "class",
          },
        ],
        "count": 1,
        "name": "Apicomplexa",
        "rank": "phylum",
      },
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [
                          {
                            "children": [
                              {
                                "count": 1,
                                "name": "Stachybotrys cylindrospora",
                                "rank": "species",
                              },
                            ],
                            "count": 1,
                            "name": "Stachybotrys",
                            "rank": "genus",
                          },
                        ],
                        "count": 1,
                        "name": "Stachybotryaceae",
                        "rank": "family",
                      },
                    ],
                    "count": 1,
                    "name": "Hypocreales",
                    "rank": "order",
                  },
                ],
                "count": 1,
                "name": "Sordariomycetes",
                "rank": "class",
              },
            ],
            "count": 1,
            "name": "Ascomycota",
            "rank": "phylum",
          },
        ],
        "count": 1,
        "name": "Fungi",
        "rank": "kingdom",
      },
    ],
    "count": 2,
    "name": "bacteria",
    "rank": "superKingdom",
  },
]
`;
